experimental/yarpl/src/yarpl/flowable/utils/SubscriptionHelper.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include "SubscriptionHelper.h"
+#include <atomic>
+#include <climits>
+
+namespace yarpl {
+namespace flowable {
+namespace internal {
+int64_t SubscriptionHelper::addCredits(
+    std::atomic<std::int64_t>* current,
+    int64_t n) {
+  for (;;) {
+    int64_t r = current->load();
+    if (r == INT64_MAX) {
+      return INT64_MAX;
+    }
+    if (n <= 0) {
+      // do nothing, return existing unmodified value
+      return r;
+    }
+    long u = r + n;
+    if (u < 0) {
+      // overflow
+      current->store(INT64_MAX);
+      return INT64_MAX;
+    }
+    // set the new number
+    if (current->compare_exchange_strong(r, u)) {
+      return u;
+    }
+    // if failed to set (concurrent modification) loop and try again
+  }
+}
+
+bool SubscriptionHelper::addCancel(std::atomic<std::int64_t>* current) {
+  for (;;) {
+    int64_t r = current->load();
+    if (r == INT64_MIN) {
+      // already cancelled
+      return false;
+    }
+    // try cancelling
+    if (current->compare_exchange_strong(r, INT64_MIN)) {
+      return true;
+    }
+    // if failed to set (concurrent modification) loop and try again
+  }
+}
+
+int64_t SubscriptionHelper::consumeCredits(
+    std::atomic<std::int64_t>* current,
+    int64_t n) {
+  for (;;) {
+    int64_t r = current->load();
+    if (n <= 0) {
+      // do nothing, return existing unmodified value
+      return r;
+    }
+    long u = r - n;
+    if (u < 0) {
+      // bad usage somewhere ... be resilient, just set to 0
```

Comment:
Doesn't assert() cause the program to terminate? That would defeat my goal of being resilient and continuing to work. 

DCHECK fails the app only in debug mode, right? I'd be okay with that.